tests/run/cpp_iterators.pyx

```diff
@@ -376,3 +376,34 @@ def test_non_built_in_reversed_function(py_v):
         vint.push_back(e)
     for e in reversed(vint):
         print(e)
+
+
+# Not strictly a C++ iterator test, but an issue with generator
+# expressions and iteration from c++ const attributes so in this
+# file as a related issue.  Mainly a test that it compiles.
+# GH-5558
+cdef extern from *:
+    """
+    struct ConstNumberHolder {
+        const int num;
+
+        explicit ConstNumberHolder(int num) :
+            num(num)
+        {}
+    };
+    """
+    cppclass ConstNumberHolder:
+        ConstNumberHolder(int num)
+
+        const int num
+
+def test_iteration_from_const_member(int num):
+    """
+    >>> test_iteration_from_const_member(5)
+    (0, 1, 2, 3, 4)
+    """
+    num_holder = new ConstNumberHolder(num)
+    try:
+        return tuple(i for i in range(num_holder.num))
```

Comment:
Might be a good idea to make `num_holder` be used inside the genexp, as well as being in the iteration expression. That way there isn't an opportunity for future optimisations to optimise out what we're testing by not including it in the scope struct. Similarly perhap pass `tuple` in as a parameter from the python code, so this can't be converted to a list comp.

Reply:
For your first suggestion: `num_holder` itself isn't included in the scope structure. It just took the `const int` type from `num_holder.num` and used that in the scope struct.

The second suggestion is a good one. It's a valid concern that the whole " `tuple` from generator expression" could be optimize out the generator expression. I'll modify it sightly I think and just return the generator expression and evaluate that in the interpreted doctest